src/EFCore.PG/Migrations/NpgsqlMigrationsSqlGenerator.cs

```diff
@@ -643,13 +643,28 @@ public virtual void Generate(NpgsqlDropDatabaseOperation operation, [CanBeNull]
 
             GenerateEnumStatements(operation, model, builder);
             GenerateRangeStatements(operation, model, builder);
-
-            foreach (var extension in operation.Npgsql().PostgresExtensions)
-                GenerateCreateExtension(extension, model, builder);
+            GenerateExtensionStatements(operation, model, builder);
 
             builder.EndCommand();
         }
 
+        #region Extension management
+
+        protected virtual void GenerateExtensionStatements(
+            [NotNull] AlterDatabaseOperation operation,
+            [NotNull] IModel model,
+            [NotNull] MigrationCommandListBuilder builder)
+        {
+            foreach (var extensionTypeToCreate in operation.Npgsql().GetPostgresExtensionsToCreate())
+                GenerateCreateExtension(extensionTypeToCreate, model, builder);
+
+            foreach (var extensionTypeToDrop in operation.Npgsql().GetPostgresExtensionsToDrop())
+                GenerateDropExtension(extensionTypeToDrop, model, builder);
+
+            foreach (var extensionTypeToAlter in operation.Npgsql().GetPostgresExtensionsToAlter())
+                GenerateAlterExtension(extensionTypeToAlter, model, builder);
+        }
```

Comment:
This method (and it's siblings for enums and ranges) could be inlined into the `Generate(AlterDatabaseOperation...)` above, but keeping them separate does leave open the possibility for downstream consumers to override just one of them.